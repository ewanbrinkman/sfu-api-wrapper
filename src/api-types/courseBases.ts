export enum SectionCode {
    Lecture = 'LEC',
    Tutorial = 'TUT',
    Lab = 'LAB',
    Seminar = 'SEM',
}

export interface RawApiCourseBase {
    title: string;
    description: string;
    corequisites: string;
    prerequisites: string;
    number: string;  // Needs to be a string, since could be "100w".
    notes: string;
    units: number;
    designation: string;
}

export interface ProcessedApiCourseBaseNotes {
    general: string;
};

export interface ProcessedApiCourseBase {
    title: string;
    description: string;
    corequisites: string;
    prerequisites: string;
    number: string;  // Needs to be a string, since could be "100w".
    notes: ProcessedApiCourseBaseNotes;
    units: number;
    designation: string;
}
