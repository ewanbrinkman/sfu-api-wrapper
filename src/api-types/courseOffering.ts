export enum DeliveryMethod {
    InPerson = 'In Person',
    DistanceEducation = 'Distance Education',
}

export enum DegreeLevel {
    Undergraduate = 'UGRD',
    Graduate = 'GRAD',
}

export enum Enrollment {
    EnrollmentSection = 'e',
    NonEnrollment = 'n',
}

export enum InstructorRole {
    PrimaryInstructor = 'PI',
    SecondaryInstructor = 'SI',
}

export interface RawApiInstructor {
    profileUrl: string;
    commonName?: string;
    firstName: string;
    lastName: string;
    phone: string;
    roleCode: InstructorRole;
    name: string;
    officeHours: string;
    office: string;
    email: string;
}

export interface ProcessedApiInstructor {
    profileUrl: string;
    commonName?: string;
    firstName: string;
    lastName: string;
    phone: string;
    role: InstructorRole;
    name: string;
    officeHours: string;
    office: string;
    email: string;
}

export enum Day {
    Monday = 'Mo',
    Tuesday = 'Tu',
    Wednesday = 'We',
    Thursday = 'Th',
    Friday = 'Fr',
    Saturday = 'Sa',
    Sunday = 'Su',
}

export enum Campus {
    Burnaby = 'Burnaby',
    Surrey = 'Surrey',
    Vancouver = 'Vancouver',
}

export enum SectionCode {
    Lecture = 'LEC',
    Tutorial = 'TUT',
    Lab = 'LAB',
    Seminar = 'SEM',
}

export interface RawApiSchedulePart {
    roomNumber: number;
    endDate: string;
    campus: Campus;
    buildingCode: string;
    days: string;
    sectionCode: SectionCode;
    startTime: string;
    isExam: boolean;
    endTime: string;
    startDate: string;
}

export interface ProcessedApiSchedulePart {
    roomNumber: number;
    endDate: string;
    campus: Campus;
    buildingCode: string;
    days: Day[];
    sectionCode: SectionCode;
    startTime: string;
    isExam: boolean;
    endTime: string;
    startDate: string;
}

export interface GradingScheme {
    description: string;
    weight: number;
}

export interface Textbook {
    details: string;
}

export interface ProcessedApiCourseOffering {
    title: string;
    name: string;
    department: string;
    number: number;
    section: string;
    units: number;
    degreeLevel: DegreeLevel;
    description: string;
    details: string;
    designation: string;
    materials: string;
    prerequisites: string;
    corequisites: string;
    requirements: string;
    educationalGoals: string;
    specialTopic: string;
    instructors: ProcessedApiInstructor[];
    deliveryMethod: DeliveryMethod;
    term: string;
    schedule: ProcessedApiSchedulePart[];
    type: Enrollment;
    gradingScheme?: GradingScheme;
    internal: {
        outlinePath: string;
        number: number;
    };
    notes: {
        general: string;
        grading: string;
        registrar: string;
        requiredReading: string;
        departmentalUndergraduateNotes: string;
        short?: string;
    };
    text: {
        required?: Textbook[];
        recommended?: Textbook[];
    };
}

export interface RawApiCourseOffering {
    info: {
        educationalGoals: string;
        notes: string;
        deliveryMethod: DeliveryMethod;
        description: string;
        section: string;
        units: number;
        title: string;
        type: Enrollment;
        classNumber: number;
        departmentalUgradNotes: string;
        prerequisites: string;
        number: number;
        requiredReadingNotes: string;
        registrarNotes: string;
        shortNote?: string;
        outlinePath: string;
        term: string;
        requirements: string;
        gradingNotes: string;
        corequisites: string;
        dept: string;
        degreeLevel: DegreeLevel;
        specialTopic: string;
        courseDetails: string;
        materials: string;
        name: string;
        designation: string;
    };
    instructor: RawApiInstructor[];
    courseSchedule: RawApiSchedulePart[];
    grades?: GradingScheme;
    requiredText?: Textbook[];
    recommendedText?: Textbook[];
}
