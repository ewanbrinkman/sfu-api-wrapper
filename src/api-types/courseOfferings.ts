import type {
    DegreeLevel,
    DeliveryMethod,
    Enrollment,
    GradingScheme,
    Textbook,
    RawApiInstructor,
    RawApiSchedulePart,
    ProcessedApiInstructor,
    ProcessedApiSchedulePart,
} from '@api-types';

export interface ProcessedApiCourseOffering {
    title: string;
    name: string;
    department: string;
    number: number;
    section: string;
    units: number;
    degreeLevel: DegreeLevel;
    description: string;
    details: string;
    designation: string;
    materials: string;
    prerequisites: string;
    corequisites: string;
    requirements: string;
    educationalGoals: string;
    specialTopic: string;
    instructors: ProcessedApiInstructor[];
    deliveryMethod: DeliveryMethod;
    term: string;
    schedule: ProcessedApiSchedulePart[];
    type: Enrollment;
    gradingScheme?: GradingScheme;
    internal: {
        outlinePath: string;
        number: number;
    };
    notes: {
        general: string;
        grading: string;
        registrar: string;
        requiredReading: string;
        departmentalUndergraduateNotes: string;
        short?: string;
    };
    text: {
        required?: Textbook[];
        recommended?: Textbook[];
    };
}

export interface RawApiCourseOffering {
    info: {
        educationalGoals: string;
        notes: string;
        deliveryMethod: DeliveryMethod;
        description: string;
        section: string;
        units: number;
        title: string;
        type: Enrollment;
        classNumber: number;
        departmentalUgradNotes: string;
        prerequisites: string;
        number: number;
        requiredReadingNotes: string;
        registrarNotes: string;
        shortNote?: string;
        outlinePath: string;
        term: string;
        requirements: string;
        gradingNotes: string;
        corequisites: string;
        dept: string;
        degreeLevel: DegreeLevel;
        specialTopic: string;
        courseDetails: string;
        materials: string;
        name: string;
        designation: string;
    };
    instructor: RawApiInstructor[];
    courseSchedule: RawApiSchedulePart[];
    grades?: GradingScheme;
    requiredText?: Textbook[];
    recommendedText?: Textbook[];
}
