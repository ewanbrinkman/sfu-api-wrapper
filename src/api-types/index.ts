export * from './requests';
export * from './courseOffering';
