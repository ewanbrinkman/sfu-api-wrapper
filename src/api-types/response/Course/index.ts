export { default as CourseData } from './processed/CourseData';
export { default as RawCourseData } from './raw/RawCourseData';
export { default as CourseSectionSummary } from './processed/CourseSectionSummary';
export { default as RawCourseSectionSummary } from './raw/RawCourseSectionSummary';
export { default as CourseType } from './common/CourseType';
export { default as SectionCode } from './common/SectionCode';
