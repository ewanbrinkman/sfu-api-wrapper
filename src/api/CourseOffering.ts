import type {
    RawApiCourseOffering,
    ProcessedApiCourseOffering,
    DegreeLevel,
    DeliveryMethod,
    Enrollment,
    GradingScheme,
    Textbook,
    ProcessedApiCourseOfferingNotes,
} from '@api-types';
import { InstructorRole } from '@api-types';
import { Instructor, SchedulePart } from '@api';

export default class CourseOffering {
    title: string;
    name: string;
    department: string;
    number: string;
    section: string;
    units: number;
    degreeLevel: DegreeLevel;
    description: string;
    details: string;
    designation: string;
    materials: string;
    prerequisites: string;
    corequisites: string;
    requirements: string;
    educationalGoals: string;
    specialTopic: string;
    instructors: Instructor[];
    deliveryMethod: DeliveryMethod;
    term: string;
    schedule: SchedulePart[];
    type: Enrollment;
    gradingScheme?: GradingScheme[];
    internal: {
        outlinePath: string;
        number: number;
    };
    notes: ProcessedApiCourseOfferingNotes;
    text: {
        required?: Textbook[];
        recommended?: Textbook[];
    };

    constructor(courseOfferingData: ProcessedApiCourseOffering) {
        Object.assign(this, courseOfferingData);
    }

    static fromRawApiCourseOffering(
        rawApiCourseOffering: RawApiCourseOffering,
    ): CourseOffering {
        const instructors: Instructor[] = rawApiCourseOffering.instructor.map(
            (rawApiInstructor) =>
                Instructor.fromRawApiInstructor(rawApiInstructor),
        );
        const schedule: SchedulePart[] =
            rawApiCourseOffering.courseSchedule.map((rawApiSchedulePart) =>
                SchedulePart.fromRawApiSchedulePart(rawApiSchedulePart),
            );

        const processedApiCourseOffering: ProcessedApiCourseOffering = {
            title: rawApiCourseOffering.info.title,
            name: rawApiCourseOffering.info.name,
            department: rawApiCourseOffering.info.dept,
            number: rawApiCourseOffering.info.number,
            section: rawApiCourseOffering.info.section,
            units: rawApiCourseOffering.info.units,
            degreeLevel: rawApiCourseOffering.info.degreeLevel,
            description: rawApiCourseOffering.info.description,
            details: rawApiCourseOffering.info.courseDetails,
            designation: rawApiCourseOffering.info.designation,
            materials: rawApiCourseOffering.info.materials,
            prerequisites: rawApiCourseOffering.info.prerequisites,
            corequisites: rawApiCourseOffering.info.corequisites,
            requirements: rawApiCourseOffering.info.requirements,
            educationalGoals: rawApiCourseOffering.info.educationalGoals,
            specialTopic: rawApiCourseOffering.info.specialTopic,
            instructors: instructors,
            deliveryMethod: rawApiCourseOffering.info.deliveryMethod,
            term: rawApiCourseOffering.info.term,
            schedule: schedule,
            type: rawApiCourseOffering.info.type,
            gradingScheme: rawApiCourseOffering.grades,
            internal: {
                outlinePath: rawApiCourseOffering.info.outlinePath,
                number: rawApiCourseOffering.info.classNumber,
            },
            notes: {
                general: rawApiCourseOffering.info.notes,
                grading: rawApiCourseOffering.info.gradingNotes,
                registrar: rawApiCourseOffering.info.registrarNotes,
                requiredReading: rawApiCourseOffering.info.requiredReadingNotes,
                departmentalUndergraduateNotes:
                    rawApiCourseOffering.info.departmentalUgradNotes,
                short: rawApiCourseOffering.info.shortNote,
            },
            text: {
                required: rawApiCourseOffering.requiredText,
                recommended: rawApiCourseOffering.recommendedText,
            },
        };
        return new CourseOffering(processedApiCourseOffering);
    }

    get primaryInstructors(): Instructor[] {
        return this.instructors.filter(
            (instructor) =>
                instructor.role === InstructorRole.PrimaryInstructor,
        );
    }

    get secondaryInstructors(): Instructor[] {
        return this.instructors.filter(
            (instructor) =>
                instructor.role === InstructorRole.SecondaryInstructor,
        );
    }
}
