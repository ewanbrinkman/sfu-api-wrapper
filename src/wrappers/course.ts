import { requestSFUAcademicCalendarApiCourses } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawApiCourse,
} from '@api-types';
import { Course } from '@api';

export default async function course(
    department: string,
    number: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Course> {
    const rawApiCourse: RawApiCourse = (
        await requestSFUAcademicCalendarApiCourses(
            year,
            term,
            department,
            number,
        )
    ).data;

    return Course.fromRawApiCourse(rawApiCourse);
}
