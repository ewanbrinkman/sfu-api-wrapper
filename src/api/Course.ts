import type {
    RawApiCourse,
    ProcessedApiCourse,
    ProcessedApiCourseBaseNotes,
    CourseOutlinesYear,
    CourseOutlinesTerm,
} from '@api-types';
import { CourseOffering, Section } from '@api';
import wrappers from '@wrappers';

export default class Course {
    title: string;
    description: string;
    corequisites: string;
    prerequisites: string;
    number: string;
    notes: ProcessedApiCourseBaseNotes;
    units: number;
    designation: string;
    recommended: string;

    department: string;
    year: CourseOutlinesYear;
    term: CourseOutlinesTerm;

    private sections: Record<string, CourseOffering | undefined>;

    constructor(
        courseData: ProcessedApiCourse,
        department: string,
        year: CourseOutlinesYear = 'current',
        term: CourseOutlinesTerm = 'current',
    ) {
        Object.assign(this, courseData);

        this.sections = courseData.sections.reduce(
            (courseOfferings, section) => {
                courseOfferings[section.section] = undefined;
                return courseOfferings;
            },
            {} as Record<string, CourseOffering | undefined>,
        );

        this.department = department;
        this.year = year;
        this.term = term;
    }

    static fromRawApiCourse(
        rawApiCourse: RawApiCourse,
        department: string,
        year: CourseOutlinesYear = 'current',
        term: CourseOutlinesTerm = 'current',
    ): Course {
        const sections: Section[] = rawApiCourse.sections.map((rawApiSection) =>
            Section.fromRawApiSection(rawApiSection),
        );

        const processedApiCourse: ProcessedApiCourse = {
            title: rawApiCourse.title,
            description: rawApiCourse.description,
            corequisites: rawApiCourse.corequisites,
            prerequisites: rawApiCourse.prerequisites,
            number: rawApiCourse.number,
            notes: {
                general: rawApiCourse.notes,
            },
            units: rawApiCourse.units,
            designation: rawApiCourse.designation,
            recommended: rawApiCourse.recommended,
            sections: sections,
        };
        return new Course(processedApiCourse, department, year, term);
    }

    public hasSection(section: string): boolean {
        return section in this.sections;
    }

    public async getSection(section: string): Promise<CourseOffering> {
        if (this.sections[section] === undefined) {
            const courseOfferingData: CourseOffering =
                await this.getCourseOffering(section);
            this.sections[section] = courseOfferingData;
        }

        return this.sections[section] as CourseOffering;
    }

    private async getCourseOffering(section: string): Promise<CourseOffering> {
        return await wrappers.courseOffering(
            this.department,
            this.number,
            section,
            this.year,
            this.term,
        );
    }

    async *[Symbol.asyncIterator](): AsyncIterableIterator<CourseOffering> {
        const sectionKeys = Object.keys(this.sections);

        for (const sectionKey of sectionKeys) {
            yield await this.getSection(sectionKey);
        }
    }

    public async getSections(): Promise<CourseOffering[]> {
        const sections: CourseOffering[] = [];

        for await (const courseOffering of this) {
            sections.push(courseOffering);
        }

        return sections;
    }
}
