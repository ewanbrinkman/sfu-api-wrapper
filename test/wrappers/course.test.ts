import course from '../../src/wrappers/course';
import Course from '../../src/api/Course';
import CourseOffering from '../../src/api/CourseOffering';
import { DegreeLevel, DeliveryMethod, Enrollment } from '../../src/api-types/courseOfferings';

describe('course', () => {
    test('request cmpt 125 fall 2022', async () => {
        const courseData: Course = await course('cmpt', '125', 2022, 'fall');

        jest.spyOn(courseData, 'sections', 'get').mockReturnValue(jest.fn().mockReturnValue({}));

        expect(courseData).toMatchSnapshot();
    });
    test('get course offerings', async () => {
        const courseData: Course = await course('cmpt', '125', 2022, 'fall');
        const courseOfferingData: CourseOffering =
            await courseData.sections['d100'];

        expect(courseOfferingData).toMatchSnapshot();
    });
});
