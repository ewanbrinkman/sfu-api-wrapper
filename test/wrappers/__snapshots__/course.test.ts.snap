// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`course get course offerings 1`] = `
CourseOffering {
  "corequisites": "",
  "degreeLevel": "UGRD",
  "deliveryMethod": "In Person",
  "department": "CMPT",
  "description": "A rigorous introduction to computing science and computer programming, suitable for students who already have some background in computing science and programming.  Intended for students who will major in computing science or a related program. Topics include: memory management; fundamental algorithms; formally analyzing the running time of algorithms; abstract data types and elementary data structures; object-oriented programming and software design; specification and program correctness; reading and writing files; debugging tools; shell commands.",
  "designation": "Quantitative",
  "details": "This course is a rigorous introduction to computing science, intended primarily for students who have already taken CMPT 120 as an introduction to algorithms and programming. Students will learn fundamental concepts of computing science, basic principles of algorithm design and software development. All case studies will be presented in C and C++. Students will also learn the basic shell commands in Linux, and will learn how to run and debug their code using VS Cod
<h2 class="ParagraphTitle">Topics</h2>
<ul>
<li>Brief review of elementary programming and problem solving; introduction to C.</li>
<li>Basic shell commands in Linux</li>
<li>Running/Debugging using Visual Studio Code</li>
<li>Performance measurements; algorithm design and analysis; asymptotics; fundamental algorithms.</li>
<li>Encodings of basic types: int, unsigned, float, char, pointer.</li>
<li>The memory model: addresses, dynamic data types, safe initialization, safe cleanup, and safe arrays.</li>
<li>Good coding style; defensive coding practices; testing.</li>
<li>Recursion: simple recursion; recursion on trees; divide and conquer algorithms.</li>
<li>Reasoning about programs: assertions, invariants, and correctness.</li>
<li>Abstract data types; information hiding;</li>
<li>Elementary data structures: stacks, queues, binary trees</li>
</ul>",
  "educationalGoals": undefined,
  "gradingScheme": undefined,
  "instructors": [
    Instructor {
      "commonName": "Igor",
      "email": "ishinkar@sfu.ca",
      "firstName": "Igor",
      "lastName": "Shinkar",
      "name": "Igor Shinkar",
      "office": "",
      "officeHours": "",
      "phone": "",
      "profileUrl": "",
      "role": "PI",
    },
  ],
  "internal": {
    "number": "5198",
    "outlinePath": "2022/fall/cmpt/125/d100",
  },
  "materials": undefined,
  "name": "CMPT 125 D100",
  "notes": {
    "departmentalUndergraduateNotes": "",
    "general": "Students with credit for CMPT 126, 129, 135 or CMPT 200 or higher may not take this course for further credit.",
    "grading": "Final Exam - 35% Lab Exam - 20% Midterm Exam - 20% Homework assignments - 25%",
    "registrar": "<p><strong>ACADEMIC INTEGRITY: YOUR WORK, YOUR SUCCESS</strong></p>
<p>SFU&rsquo;s Academic Integrity website <a href="http://www.sfu.ca/students/academicintegrity.html">http://www.sfu.ca/students/academicintegrity.html</a> is filled with information on what is meant by academic dishonesty, where you can find resources to help with your studies and the consequences of cheating. Check out the site for more information and videos that help explain the issues in plain English.</p>
<p>Each student is responsible for his or her conduct as it affects the university community. Academic dishonesty, in whatever form, is ultimately destructive of the values of the university. Furthermore, it is unfair and discouraging to the majority of students who pursue their studies honestly. Scholarly integrity is required of all members of the university. <a href="http://www.sfu.ca/policies/gazette/student/s10-01.html">http://www.sfu.ca/policies/gazette/student/s10-01.html</a></p>",
    "requiredReading": "Your personalized Course Material list, including digital and physical textbooks, are available through the SFU Bookstore website by simply entering your Computing ID at: <a href="https://shop.sfu.ca/course-materials/my-personalized-course-materials">shop.sfu.ca/course-materials/my-personalized-course-materials</a>.",
    "short": undefined,
  },
  "number": "125",
  "prerequisites": "CMPT 120 or CMPT 130, with a minimum grade of C-.",
  "requirements": undefined,
  "schedule": [
    SchedulePart {
      "campus": "Burnaby",
      "days": [
        "Mo",
      ],
      "endDate": "Tue Dec 06 00:00:00 PST 2022",
      "endTime": "Tue Dec 06 00:00:00 PST 2022",
      "isExam": false,
      "sectionCode": "LEC",
      "startDate": "14:30",
      "startTime": "14:30",
    },
    SchedulePart {
      "campus": "Burnaby",
      "days": [
        "We",
      ],
      "endDate": "Tue Dec 06 00:00:00 PST 2022",
      "endTime": "Tue Dec 06 00:00:00 PST 2022",
      "isExam": false,
      "sectionCode": "LEC",
      "startDate": "14:30",
      "startTime": "14:30",
    },
  ],
  "section": "D100",
  "specialTopic": "",
  "term": "Fall 2022",
  "text": {
    "recommended": undefined,
    "required": undefined,
  },
  "title": "Introduction to Computing Science and Programming II",
  "type": "e",
  "units": "3",
}
`;

exports[`course request cmpt 125 fall 2022 1`] = `
Course {
  "corequisites": "",
  "department": "cmpt",
  "description": "A rigorous introduction to computing science and computer programming, suitable for students who already have some background in computing science and programming. Intended for students who will major in computing science or a related program. Topics include: memory management; fundamental algorithms; formally analyzing the running time of algorithms; abstract data types and elementary data structures; object-oriented programming and software design; specification and program correctness; reading and writing files; debugging tools; shell commands. Prerequisite: CMPT 120 or CMPT 130, with a minimum grade of C-. Students with credit for CMPT 126, 129, 135 or CMPT 200 or higher may not take this course for further credit. Quantitative.",
  "designation": "Quantitative",
  "notes": {
    "general": "Students with credit for CMPT 126, 129, 135 or CMPT 200 or higher may not take this course for further credit.",
  },
  "number": "125",
  "prerequisites": "Prerequisite: CMPT 120 or CMPT 130, with a minimum grade of C-.",
  "recommended": "",
  "sections": {
    "d100": undefined,
    "d101": undefined,
    "d102": undefined,
    "d103": undefined,
    "d104": undefined,
    "d105": undefined,
    "d106": undefined,
    "d107": undefined,
    "d108": undefined,
  },
  "term": "fall",
  "title": "Introduction to Computing Science and Programming II",
  "units": "3",
  "year": 2022,
}
`;
